Add global phase -1, relative phase i, and complex relative phase tasks to Single-Qubit Gates kata (#1369)
microsoft/qsharp: katas/content/single_qubit_gates/global_phase_change/solution.md

We recognize that a global phase change can be accomplished by using the R rotation gate with the PauliI (identity) gate.
As a reminder, the R gate is defined as $R_{\mu}(\theta) = \exp(\frac{\theta}{2}i\cdot\sigma_{\mu})$, where $\sigma_{\mu}$ is one of the Pauli gates I, X, Y or Z.

> Note that a global phase is not detectable and has no physical meaning - it disappears when you take a measurement of the state.  

> > You can read more about this in the [Single-qubit measurements tutorial](../tutorials/SingleQubitSystemMeasurements/SingleQubitSystemMeasurements.ipynb#Measurements-in-arbitrary-orthogonal-bases).

Reviewer: Old links don't work in the new website
```suggestion
> You can read more about this in the Single-Qubit Measurements kata.
```

For the problem at hand, we'll use the rotation gate $R_{\mu}(\theta) = \exp(\frac{\theta}{2}i\cdot\sigma_{\mu})$ with $\sigma_{\mu} = I$.

> $R(PauliI, 2\pi) = \exp(\frac{2\pi}{2} iI) = \exp(i\pi) I = (\cos\pi + i\sin\pi) I$ and, since $\cos\pi = -1$ and $\sin\pi = 0$, we have that $R(PauliI, 2\pi) = -I$:

Reviewer: This long line does not render as well in the narrower environment that we have, let's rephrase
```suggestion
$$R(PauliI, 2\pi) = \exp(\frac{2\pi}{2} iI) = \exp(i\pi) I = (\cos\pi + i\sin\pi) I$$
Since $\cos\pi = -1$ and $\sin\pi = 0$, we have that $R(PauliI, 2\pi) = -I$:
```

$$
R(\beta |0\rangle + \gamma |1\rangle) =
 -1\begin{bmatrix}1 & 0 \\\ 0 & 1 \end{bmatrix}
 \begin{bmatrix}\beta\\\ \gamma\\\ \end{bmatrix}=
 \begin{bmatrix}-1 & 0 \\\ 0 & -1 \end{bmatrix}
 \begin{bmatrix}\beta\\\ \gamma\\\ \end{bmatrix} =

>  \begin{bmatrix}-1\cdot\beta + 0\cdot\gamma\\\ 0\cdot\beta +  -1\cdot\gamma \\\ \end{bmatrix}=

Reviewer: ```suggestion
 \begin{bmatrix}(-1)\cdot\beta + 0\cdot\gamma\\\ 0\cdot\beta +  (-1)\cdot\gamma \\\ \end{bmatrix}=
```

\begin{bmatrix}-\beta\\\ -\gamma\\\ \end{bmatrix}=

> - \beta |0\rangle - \gamma |1\rangle

Reviewer: This formula doesn't render, I tracked it down to this line being interpreted as a bullet list instead of a formula, the fix is 
```suggestion
-\beta |0\rangle - \gamma |1\rangle
```

$$

> The test harness for this test shows the result of applying the _controlled_ variant of your solution to be able to detect the phase change.

Reviewer: We're using the same testing harness here as in other tasks, so let's remove this note
```suggestion
```

Author: Got it.

Our task is to implement an operation that transforms the given qubit state from $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$ to $- \beta |0\rangle - \gamma |1\rangle$.

To do that, we utilize a sequence of gates. The Pauli Z gate followed by the Pauli X gate can be used to achieve this effect when applied in succession twice.

1. **Apply the Pauli Z gate**: The Z gate multiplies the $|1\rangle$ state by $-1$ but does not change the $|0\rangle$ state, converting our state $\beta |0\rangle + \gamma |1\rangle$ to $\beta |0\rangle - \gamma |1\rangle$.

   The matrix representation of the Z gate is:

   $$
   Z =
   \begin{bmatrix}1 & 0 \\\ 0 & -1 \end{bmatrix}
   $$

2. **Apply the Pauli X gate**: The X gate flips the $|0\rangle$ and $|1\rangle$ basis states, converting our $\beta |0\rangle - \gamma |1\rangle$ state to $\beta |1\rangle - \gamma |0\rangle$.

   The matrix representation of the X gate is:

   $$
   X =
   \begin{bmatrix}0 & 1 \\\1 & 0\end{bmatrix}
   $$

> 3. **Repeat the Z and X gates**: Applying the Z gate again will multiply the $|1\rangle$ state (that used to be $|0\rangle$), converting our state $\beta |1\rangle - \gamma |0\rangle$ to $\beta |1\rangle - \gamma |0\rangle$. Finally, the second X gate will restore the original basis states, but now with both amplitudes having acquired an additional phase of $-1$. This means our state has been multiplied by $-1$, achieving the required global phase change.

Reviewer: Missed one sign :-(
```suggestion
3. **Repeat the Z and X gates**: Applying the Z gate again will multiply the $|1\rangle$ state (that used to be $|0\rangle$), converting our state $\beta |1\rangle - \gamma |0\rangle$ to $- \beta |1\rangle - \gamma |0\rangle$. Finally, the second X gate will restore the original basis states, but now with both amplitudes having acquired an additional phase of $-1$. This means our state has been multiplied by $-1$, achieving the required global phase change.
```
